google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb

```diff
@@ -60,6 +60,17 @@ def modify_ack_deadline deadline, ack_ids
           end
           alias delay modify_ack_deadline
 
+          def renew_lease deadline, ack_ids
+            return if ack_ids.empty?
+
+            ack_ids.each do |ack_id|
+              # Do not overwrite pending actions when renewing leased messages.
+              @register[ack_id] ||= deadline
```

Comment:
This is shorthand for `@register[ack_id] = @register[ack_id] || deadline` which still leaves room for concurrent modification. The scenario is, someone else sets it to `:ack` between when we read the old value and set the new one, and as a result we are able to overwrite an `:ack`.

`ConcurrentHash` is just a Ruby hash and does not provide an atomic check-and-set operation, so introducing this logic means we now need to synchronize around access to `@register`.

Reply:
The call to `Stream#renew_lease!` already does `synchronize` around the subscriber (including the subscriber's buffer) when it calls this method which modifies @register. I believe we are good here, right?

Would you prefer we switch `@register` to a normal Ruby Hash because of this?

Reply:
_Anything_ that modifies `@register` (i.e. anything that calls `TimedUnaryBuffer#acknowledge` or `TimedUnaryBuffer#modify_ack_deadline`) would also need to be synchronized, and _on the same monitor_. So it's safer for the monitor to be local to this class than depend on all callers to do the right thing.

And... yeah, I'd prefer a normal Ruby hash. (Incidentally, it turns out that, in MRI, `Concurrent::Hash` is just aliased to `::Hash` anyway. They depend on the observation that all the code that matters happens to be implemented in C and is protected by the GIL.)

Reply:
Okay, I have double checked all calls to the `buffer` methods to make sure they all accessed within a `synchronize` block. I have updated `@register` to use a normal Ruby Hash. And I have removed the method aliases from the internal methods, so it should be easier to read now. PTAL and let me know what you think.

Reply:
Ok. Guess I didn't remember that the usage was limited at this point.